src/crimson/os/seastore/cache.h

```diff
@@ -181,7 +181,7 @@ class Cache {
       auto ret = TCachedExtentRef<T>(static_cast<T*>(&*iter));
       return ret->wait_io(
       ).then([ret=std::move(ret)]() mutable -> get_extent_ret<T> {
-	if (!ret->is_retired()) {
+	if (ret->is_valid()) {
```

Comment:
`!is_valid()` include both RETIRED and INVALID scenarios:
https://github.com/ceph/ceph/blob/7359eca5d07170124155dd948fc7443245bd94a7/src/crimson/os/seastore/cached_extent.h#L226-L229

I think logically we only return eagain due to transaction confliction, not when we try to read a retired extent, which may lead to endless loop.

Reply:
here need check valid,  not invalid.

Reply:
valid means sate not RETIRED and INVALID, then the extent can return. the original only check state not RETIRED, missed INVALID state.

Reply:
Yeah, I actually meant to comment on the `else` portion when `is_valie() == false`.

Seem to me reading a retired extent (could be due to user's mistake) should not return eagain.

Reply:
Reading an extent retired within the same transaction is I think always a bug.

Reply:
I agree, maybe `ceph_abort()` this case, but not eagain.

Reply:
@athanatos what do you think retired extent, return eagain or ceph_abort()?

Reply:
@cyx1231st why do you think eagain for RETIRED extent will cause endless loop?  when the confliction transaction finish, the caller who get eagain redo will get right extent, right?

Reply:
If a user tries to read a retired extent in its buggy logic and got eagain, probably it will retry and hit the same eagain.

Reply:
ceph_abort will be fine -- I think you can only see a retired extent here due to a bug.

Reply:
@athanatos I do meet retired here. are you sure it is a bug and not a confliction which need eagain return?

Reply:
@athanatos if it is a bug, I will try to fix it in another PR.

Reply:
> ceph_abort will be fine -- I think you can only see a retired extent here due to a bug.

@athanatos Hi, I do meet a retired extent here, I think this may not be a bug. It seems that retiring an extent would drop it from the cache, and `Cache::get_extent` will try to find a CachedExtent with the specific offset. It looks to me that this would inevitably lead to scenarios where we would meet a retired extent here. So I think maybe we should allow this case, am I right? Thanks:-)

Reply:
> ceph_abort will be fine -- I think you can only see a retired extent here due to a bug.

@athanatos I do meet a retired extent here, it seems that, when rewriting a block, the old one would get retired but not replaced, so there are chances where a retired extent is met here, I think to correct it, we should replace the old extent with the new one when rewriting that block, am I right? thanks.

Reply:
see https://github.com/ceph/ceph/pull/41801